plugins/CoreAdminHome/Commands/InvalidateReportData.php

```diff
@@ -0,0 +1,272 @@
+
+        $datesByPeriod = $this->getPeriodsToInvalidateFor($periodTypes, $dateRanges, $cascade);
+
+        foreach ($datesByPeriod as $periodType => $dates) {
+            $output->writeln("Invalidating $periodType periods...");
+
+            if ($output->getVerbosity() > OutputInterface::VERBOSITY_NORMAL) {
+                $output->writeln("  [ dates = " . implode(', ', $dates) . " ]");
+            }
+
+            if ($dryRun) {
+                $output->writeln("[Dry-run] invalidating archives for site = [ " . implode(', ', $sites)
+                    . " ], dates = [ " . implode(', ', $dates) . " ], period = [ $periodType ]");
+            } else {
+                $invalidationResult = $invalidator->markArchivesAsInvalidated($sites, $dates, $periodType);
+
+                if ($output->getVerbosity() > OutputInterface::VERBOSITY_NORMAL) {
+                    $output->writeln($invalidationResult->makeOutputLogs());
+                }
+            }
+        }
+    }
+
+    private function getSitesToInvalidateFor(InputInterface $input)
+    {
+        $sites = $input->getOption('sites');
+
+        $siteIds = Site::getIdSitesFromIdSitesString($sites);
+        if (empty($siteIds)) {
+            throw new \InvalidArgumentException("Invalid --sites value: '$sites'.");
+        }
+
+        $allSiteIds = SitesManagerAPI::getInstance()->getAllSitesId();
+        foreach ($siteIds as $idSite) {
+            if (!in_array($idSite, $allSiteIds)) {
+                throw new \InvalidArgumentException("Invalid --sites value: '$sites', there are no sites with IDs = $idSite");
+            }
+        }
+
+        return $siteIds;
+    }
+
+    private function getPeriodTypesToInvalidateFor(InputInterface $input)
+    {
+        $periods = $input->getOption('periods');
+        if (empty($periods)) {
+            throw new \InvalidArgumentException("The --periods argument is required.");
+        }
+
+        if ($periods == self::ALL_OPTION_VALUE) {
+            $result = array_keys(Piwik::$idPeriods);
+            unset($result[4]); // remove 'range' period
+            return $result;
+        }
+
+        $periods = explode(',', $periods);
+        $periods = array_map('trim', $periods);
+
+        foreach ($periods as $periodIdentifier) {
+            if ($periodIdentifier == 'range') {
+                throw new \InvalidArgumentException(
+                    "Invalid period type: invalidating range periods is not currently supported.");
+            }
+
+            if (!isset(Piwik::$idPeriods[$periodIdentifier])) {
+                throw new \InvalidArgumentException("Invalid period type '$periodIdentifier' supplied in --periods.");
+            }
+        }
+
+        return $periods;
+    }
+
+    /**
+     * @param InputInterface $input
+     * @return Date[][]
+     */
+    private function getDateRangesToInvalidateFor(InputInterface $input)
+    {
+        $dateRanges = $input->getOption('dates');
+        if (empty($dateRanges)) {
```

Comment:
The code starting from here could be kinda moved into a public method in another class as well. It would be could to have separate, small tests for this one I think as I would for example not feel confident to change it without "fear" or a certain "risk". I know it's tested by the command itself but not sure if really each case it goes is tested